### ISSUE TITLE
Farm may be useful/interesting

### ISSUE THREAD

**Issue by pvh** [dweb-camp-2019/projects#1]
We built `farm`, a [p2p collaborative web development environment](https://github.com/inkandswitch/farm) that builds offline-first real-time collaborative apps. It might be useful to the project and doesn't require internet but *is* an electron app you'll need to install.

It could be cool for building ad-hoc programs to organize people, capture notes, depict the camp, or pretty much anything else you could imagine you'd want real-time Docs-like collaboration for.

It uses discovery-swarm and can do peer discovery over mDNS & via the dat community's DNS-SD and using it requires installing an electron app as of this writing. 

As of March it's transitioning from a research project to usable software and may-or-may-not actually work in the field. If someone wants it to, that increases the likelihood of us fixing the bugs.

Feel free to reach out to me here if you have questions about it.

If folks can find each other via mDNS, this'd be pretty easy, but I'm guessing the network will be too big for that to work?

**Comment by benhylau** [comment]
@amark adding to @RangerMauve's answer, what I was referring to is purely for discovery purposes. I understand that once peered with these links [ A-B, B-C, B-D, D-E, D-F ] messages will route across the network even if A-F doesn't exist. But when we spin up A and B, how do they form A-B in the first place?

If it's on the same switch, perhaps mdns peer discovery would allow them to automatically find each other and form A-B. However, if they are on two different switches, that's as if the A if plugged into my home, and B plugged into my office, behind their own NATs. How does GUN handle discovery / peering in this scenario?

**Comment by amark** [comment]
@benhylau @RangerMauve exactly, two laptops on my same home wireless router discover each other and start syncing.

But, IDK how this works with different WiFi routers - my friends' place with multiple routers, the repeater is still a different network that has to be connected to separately, and I have no clue what happens with multicast in those settings (???) can somebody inform/educate me?

My stupid naive thought is: Can a device plugged into A also wirelessly connect to B, and vice versa, a device plugged into B wirelessly connect with A? (I assume this isn't necessary) Now device has ethernet to A's WiFi connected peers, and is WiFi connected to B.

 :P :P :P do OSes even let me use Ethernet + WiFi simultaneously? They should!!! I don't think I've ever seen/done this in actuality tho, and I honestly don't play around with this stuff much. But clearly + obviously need to and should.

Please teach me or link me to "WiFi Mesh Networks for Dummies 101" videos! :D :P

Assuming that WiFi A and WiFi B can also directly connect to each other, aren't they still going to be on some sort of same subnet that multicast would broadcast across?

**Comment by amark** [comment]
home <-> office is just usually IP addresses if available. If not, leaving it to WebRTC to NAT traversal and crossing fingers it'll work (WebRTC fails lots for me, only like 40% success rate). Final fallback is through websocket relay peer(s) with known publicly exposed IP/DNS addresses, but that is lame when if we have our own mesh network can't we be doing multicast!??

**Comment by benhylau** [comment]
@amark I feel like we are squatting the Farm thread :D but let me try to address the open questions, then move back to #2 for continued discussions heh

@benhylau @RangerMauve exactly, two laptops on my same home wireless router discover each other and start syncing.

But, IDK how this works with different WiFi routers - my friends' place with multiple routers, the repeater is still a different network that has to be connected to separately, and I have no clue what happens with multicast in those settings (???) can somebody inform/educate me?

My stupid naive thought is: Can a device plugged into A also wirelessly connect to B, and vice versa, a device plugged into B wirelessly connect with A? (I assume this isn't necessary) Now device has ethernet to A's WiFi connected peers, and is WiFi connected to B.

>do OSes even let me use Ethernet + WiFi simultaneously?

Yup, you can have many many interfaces, they can be connected to the same or different networks.  Which path traffic flows through will be determined by your routing table (e.g. `route -n`). If you bridge the interfaces, they'd behave as if they are plugged in to the same network switch.

>Can a device plugged into A also wirelessly connect to B, and vice versa, a device plugged into B wirelessly connect with A? (I assume this isn't necessary) Now device has ethernet to A's WiFi connected peers, and is WiFi connected to B.
>[ . . . ]
>Assuming that WiFi A and WiFi B can also directly connect to each other, aren't they still going to be on some sort of same subnet that multicast would broadcast across?

I _think_ if the device that is plugged into A and wirelessly connected to B **has its two network interfaces bridged** (e.g. eth0 and wlan0), then it will let the multicast flow through. Not 100% sure.

>my friends' place with multiple routers

Depends on topology, I think if one is nested in the LAN of another then it will get all the broadcast.

>repeater is still a different network

Depends on implementation of the repeater, they have all kinds of stuff out there!

**Comment by benhylau** [comment]
@pvh I want to loop back on the original discussion of this thread, and see if farm will be leading sessions at Camp. I don't think it's contingent on having ready / reliable software, perhaps this is exactly the right time to hold discussion sessions that could inform what the software will be!
